src/lib/massa-react/utils/operationHandler.ts

```diff
@@ -0,0 +1,99 @@
+import { Operation, OperationStatus } from '@massalabs/massa-web3';
+import { toast } from '../../../components';
+import { ERROR_STATUSES } from '../hooks/const';
+import { ToasterMessage } from '../hooks/types';
+import { showToast, logSmartContractEvents } from '.';
+
+export type OperationState = {
+  isOpPending: boolean;
+  isPending: boolean;
+  isSuccess: boolean;
+  isError: boolean;
+  opId?: string;
+};
+
+export async function executeOperation(
+  operation: Operation,
+  messages: ToasterMessage,
+  final: boolean,
+  isMainnet: boolean | undefined,
+  setState: React.Dispatch<React.SetStateAction<OperationState>>,
+): Promise<void> {
+  try {
+    setState((prev) => ({
+      ...prev,
+      opId: operation.id,
+    }));
+
+    const loadingToastId = showToast(
+      'loading',
+      messages.pending,
+      operation.id,
+      isMainnet,
+    );
+
+    const finalStatus = final
+      ? await operation.waitFinalExecution()
+      : await operation.waitSpeculativeExecution();
+
+    dismissLoadingToast(loadingToastId);
+
+    if (finalStatus === OperationStatus.NotFound) {
+      handleOperationTimeout(messages.timeout, operation.id, setState);
+      throw new Error('Operation not found');
+    } else if (ERROR_STATUSES.includes(finalStatus)) {
+      logSmartContractEvents(operation.provider, operation.id);
+      throw new Error(`Operation failed with status: ${finalStatus}`);
+    } else {
+      handleOperationSuccess(messages.success, operation.id, setState);
+    }
+  } catch (error) {
+    if (error instanceof Error) {
+      handleOperationError(error, messages.error, operation.id, setState);
+    }
+    throw error;
+  } finally {
+    setState((prev) => ({
+      ...prev,
+      isOpPending: false,
+      isPending: false,
+    }));
```

Comment:
Could make more easy to read if wrapped like this:

```
updateOpState(setState, { isOpPending: false, isPending: false });

with:
function updateOpState<State>(
  setState: React.Dispatch<React.SetStateAction<State>>,
  partialState: Partial<State>
) {
  setState((prevState) => ({
    ...prevState,
    ...partialState,
  }));
}

Reply:
You're right thanks